src/main/java/WebApp/Enterprise/Pollima/controller/CompanyController.java

```diff
@@ -26,23 +27,57 @@ public class CompanyController {
     private CompanyService companyService;
 
     @GetMapping
-    public String company(Model model){
-        List<Company> companyList = companyService.findAll();
-        model.addAttribute("CompanyList",companyList);
-        model.addAttribute("AddNewCompany", CompanyForm.builder().build());
+    public String company(Model model, @PageableDefault(size = 2) Pageable pageable) {
+        Page<Company> companyPage = companyService.findAll(pageable);
+        List<Company> companyList = companyPage.getContent();
+        model.addAttribute("companyList", companyList);
+        model.addAttribute("saveNewCompany", CompanyForm.builder().build());
+        model.addAttribute("companyPage", companyPage);
         return "company";
     }
 
+
     @PostMapping
-    private String saveCompany(@ModelAttribute("saveNewCompany") CompanyForm form){
+    private String saveCompany(@Valid @ModelAttribute("saveNewCompany") CompanyForm form, BindingResult result){
+
+        if (form.getId()==null) {
+            companyService.findByCompanyName(form.getCompanyName()).ifPresent(company ->
+                    result.rejectValue("companyName", "company.exist", new Object[]{form.getCompanyName()}, "company.exist")
+            );
+        }
+
+        if (result.hasErrors()){
+            return "company";
+        }
+
         companyService.save(Company.builder()
-                .ID(form.getID())
-                .CompanyName(form.getCompanyName())
+                .id(form.getId())
+                .companyName(form.getCompanyName())
                 .contactPerson(form.getContactPerson())
                 .office(form.getOffice())
                 .contactNo(form.getContactNo())
                 .build());
         return "redirect:/company";
     }
 
+    @GetMapping("/delete")
+    private String deleteCompany(@RequestParam(name = "id") long id){
+        companyService.delete(id);
```

Comment:
Do not delete by id. Always delete by object. Check if the object exists first and then delete by the object.
Something like `companyService.findById(id).ifPresent(company -> {`